src/core/transport/chttp2_transport.c

```diff
@@ -823,24 +823,26 @@ static void unlock(transport *t) {
   finish_reads(t);
 
   /* gather any callbacks that need to be made */
-  if (!t->calling_back && cb) {
+  if (!t->calling_back) {
     perform_callbacks = prepare_callbacks(t);
     if (perform_callbacks) {
       t->calling_back = 1;
     }
-    if (t->error_state == ERROR_STATE_SEEN && !t->writing) {
-      call_closed = 1;
-      t->calling_back = 1;
-      t->cb = NULL; /* no more callbacks */
-      t->error_state = ERROR_STATE_NOTIFIED;
-    }
-    if (t->num_pending_goaways) {
-      goaways = t->pending_goaways;
-      num_goaways = t->num_pending_goaways;
-      t->pending_goaways = NULL;
-      t->num_pending_goaways = 0;
-      t->cap_pending_goaways = 0;
-      t->calling_back = 1;
+    if (cb) {
+      if (t->error_state == ERROR_STATE_SEEN && !t->writing && !t->calling_back) {
```

Comment:
Want to merge and fix Travis anyway... This will only add some latency
disconnecting. I'll fix it up better this morning.

On Thu, May 7, 2015, 7:34 AM Nicolas Noble <notifications@github.com> wrote:

> In src/core/transport/chttp2_transport.c
> <https://github.com/grpc/grpc/pull/1499#discussion_r29857302>:
>
> >      perform_callbacks = prepare_callbacks(t);
> >      if (perform_callbacks) {
> >        t->calling_back = 1;
> >      }
> > -    if (t->error_state == ERROR_STATE_SEEN && !t->writing) {
> > -      call_closed = 1;
> > -      t->calling_back = 1;
> > -      t->cb = NULL; /* no more callbacks */
> > -      t->error_state = ERROR_STATE_NOTIFIED;
> > -    }
> > -    if (t->num_pending_goaways) {
> > -      goaways = t->pending_goaways;
> > -      num_goaways = t->num_pending_goaways;
> > -      t->pending_goaways = NULL;
> > -      t->num_pending_goaways = 0;
> > -      t->cap_pending_goaways = 0;
> > -      t->calling_back = 1;
> > +    if (cb) {
> > +      if (t->error_state == ERROR_STATE_SEEN && !t->writing && !t->calling_back) {
>
> I was kind of wondering that, yes :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc/pull/1499/files#r29857302>.
>